src/dmd/target.d

```diff
@@ -728,6 +728,36 @@ struct Target
             return false;
         }
     }
+
+    /***
+     * Determine the size a value of type `t` will be when it
+     * is passed on the function parameter stack.
+     * Params:
+     *  loc = location to use for error messages
+     *  t = type of parameter
+     * Returns:
+     *  size used on parameter stack
+     */
+    extern (C++) ulong parameterSize(const ref Loc loc, Type t)
+    {
+        if (!global.params.is64bit &&
+            (global.params.isFreeBSD || global.params.isOSX))
+        {
+            /* These platforms use clang, which regards a struct
+             * with size 0 as being of size 0 on the parameter stack,
+             * even while sizeof(struct) is 1.
+             * It's an ABI incompatibility with gcc.
+             */
+            if (t.ty == Tstruct)
```

Comment:
Are you sure this is correct for both C and C++?

I ask because gcc and g++ has a discrepancy between empty aggregates. One reports sizeof == 0 while the other reports 1.

Reply:
You're right, gcc and g++ do behave differently. gcc, clang, clang++ have 0 sized structs on the parameter stack taking up no space, g++ they take up space.

Reply:
With g++, `extern "C" { ... }` has no effect on the empty struct size, it stays at 1. So g++ cannot interface with code compiled with gcc. (!)

Reply:
Which leads us to a hopeless problem - we cannot be compatible with both g++ in `extern "C"` and with gcc.

Reply:
clang++ should give you a non-zero sized struct in cpp code.
If it does not it has a bug.

Reply:
> You're right, gcc and g++ do behave differently. gcc, clang, clang++ have 0 sized structs on the parameter stack taking up no space, g++ they take up space.

As per Stefan's comment.  The distinction here is C vs. C++.  In C empty structs have no size, but in C++, the standard dictates that an empty struct is equivalent to `struct S { char pad; }`.

Citation:
https://gcc.gnu.org/onlinedocs/gcc-7.3.0/gcc/Empty-Structures.html#Empty-Structures
http://www.stroustrup.com/bs_faq2.html#sizeof-empty

Reply:
Now, as per 14.1.3 of the spec (https://dlang.org/spec/struct.html#struct_layout)

> The default layout of the fields of a struct is an exact match with the associated C compiler.

Ignoring the phrasing of the word "associated C compiler" and taking that to just mean "D structs are C ABI compatible".

That means that all structs should have a size `0`, therefore we can never be compatible with C++.  But you can however define `struct MyCppStruct { char pad; }` to mimic the equivalent behaviour.

Reply:
The clang++ equivalent behavior cannot be mimicked by adding a dummy field. clang++ returns sizeof==1 for empty structs, but they use 0 size on the parameter stack.

Reply:
The whole thing is a mess.

Reply:
Or at least perhaps gdc will be g++ compatible, ldc clang++ compatible. You're just stuck in the middle.